src/main/kotlin/org/pkl/lsp/Builder.kt

```diff
@@ -34,12 +34,48 @@ import org.pkl.lsp.ast.PklModule
 import org.pkl.lsp.ast.PklModuleImpl
 import org.pkl.lsp.ast.PklNode
 import org.pkl.lsp.ast.Span
+import org.pkl.lsp.services.PackageEventType
+import org.pkl.lsp.services.ProjectEventType
+import org.pkl.lsp.services.packageTopic
+import org.pkl.lsp.services.projectTopic
 
 class Builder(private val server: PklLSPServer, project: Project) : Component(project) {
   private val runningBuild: MutableMap<String, CompletableFuture<PklModule?>> = mutableMapOf()
 
+  private val buildCache: MutableMap<URI, PklModule> = ConcurrentHashMap()
```

Comment:
Some food for thought: I think there's some room for improvement here. Right now, the builder is responsible for both parsing source files into `PklModule`, and also for running diagnostics.

I'm thinking we can have `interface VirtualFile { val pklModule: CompletableFuture<PklModule> }`, where that property gets invalidated whenever the file changes. Since it's now a modification tracker, it should be pretty simple to manage (e.g. we can use `CachedValuesManager` for this).

Then, we just need a diagnostics manager that listens for file change events, and requests new diagnostics, and we should be able to eliminate `Builder`.